### PR TITLE
Add Huawei Ascend NPU training support

### PR DESCRIPTION
**Pull request by glenn-jocher**
## Summary

- add validated single- and multi-NPU Ascend training through the existing device, trainer, validator, dataloader, AMP, memory, profiling, and AutoBatch owners
- derive accelerator behavior from the selected `torch.device` through PyTorch native device-module mapping, retaining the PyTorch 1.8 fallback and CUDA-first default while accommodating Ascend `npu`, Intel `xpu`, and AMD ROCm standard `cuda` device type
- bind distributed ranks locally and select NCCL/Gloo, HCCL, or XCCL for CUDA, Ascend, or Intel devices
- document Ascend among the standard training options and connect training to the existing Ascend export/deployment guide
- add the missing AMD integration guide with supported ROCm training syntax and explicit MIGraphX, DirectML, and Ryzen AI NPU boundaries
- bump the package patch version to 8.4.111

The implementation does not add an Ascend-specific trainer or parallel abstraction, and existing GPU-facing labels remain unchanged.

## Validation

- `python -m pytest tests/test_engine.py tests/test_python.py -q` — 167 passed, 2 skipped
- focused device, dataloader, PyTorch 2.6/2.7 pinning, profiling, NMS, XCCL, MPS compatibility, and post-merge checks
- Ruff format/check on all changed Python files
- generated API references and Prettier 3.8.5 docs formatting
- full MkDocs build, including the new AMD page
- `git diff --check`
- exact-head CI: all platform tests, benchmarks, GPU, isolated exports, Python 3.8/PyTorch 1.8 floor, docs, Actions, CodeQL, CLA, and patch coverage passed
- cold full-diff GPT-5.6 Sol xhigh review and same-reviewer deltas — LGTM
- Claude Code Opus 5 xhigh full/delta review — LGTM

## Hardware validation

No Ascend NPU or Intel XPU is attached to the development host. Single-NPU AMP train/validation, two-NPU HCCL, Intel XPU, and XCCL runtime smokes remain physical-hardware validation items.

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary

🚀 Adds broad PyTorch backend support for AMD ROCm and Huawei Ascend NPUs, while improving device-agnostic training, validation, profiling, and memory management across Ultralytics.

### 📊 Key Changes

- 🟢 Adds a comprehensive [AMD ROCm integration guide](../integrations/amd.md) covering:
  - Training, validation, inference, and multi-GPU execution on supported AMD GPUs.
  - Standard `device=0` and `device=cuda:0` usage through PyTorch’s ROCm compatibility layer.
  - Current support status for MIGraphX, DirectML, Ryzen AI NPU, and ONNX runtimes.
- 🧠 Expands Huawei Ascend documentation with single- and multi-NPU training through `torch_npu`, AMP, checkpointing, AutoBatch, and HCCL-based distributed training.
- ⚙️ Refactors device handling through `get_torch_device_backend()` to support CUDA, ROCm, Ascend NPU, and Intel XPU backends.
- 🔄 Updates distributed training to use backend-specific device setup and communication:
  - HCCL for Ascend NPUs.
  - XCCL for Intel XPUs.
  - NCCL or Gloo for CUDA-compatible devices.
- 📦 Makes dataloaders, AMP, memory monitoring, AutoBatch, profiling, synchronization, and non-blocking transfers accelerator-aware.
- 🛡️ Prevents unsupported torchvision NMS kernels from being used on Ascend NPU and Intel XPU devices.
- 📝 Updates device configuration, training arguments, integration navigation, API references, and ONNX documentation.
- 🔖 Bumps the package version from `8.4.110` to `8.4.111`.

### 🎯 Purpose & Impact

- 🌍 Enables more users to train and run Ultralytics models on non-NVIDIA hardware, especially AMD GPUs and Huawei Ascend NPUs.
- 🚄 Improves compatibility and performance for multi-device training across ROCm, Ascend, Intel XPU, and CUDA environments.
- 🧩 Reduces backend-specific code duplication by routing accelerator operations through the appropriate PyTorch device module.
- ✅ Provides clearer installation guidance and more accurate expectations around AMD technologies that are not yet natively supported.
- 🔧 Improves reliability for AMP, memory handling, profiling, AutoBatch, and distributed execution on alternative accelerators.
- ⚠️ Existing CUDA workflows remain supported, while users of AMD ROCm or Ascend hardware must install compatible PyTorch and vendor runtime packages.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @glenn-jocher, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This automated checklist is here to help ensure a seamless integration. An Ultralytics engineer will assist with review, so please check the following:
- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_ — Bruce Lee
For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

The device parsing, rank binding, HCCL setup, memory paths, and NPU batch-transfer changes are well integrated with the existing owners. Two gaps remain: AMP is broken for supported pre-2.6 PyTorch/torch_npu combinations, and NPU profiling/validation speed reporting does not synchronize asynchronous NPU work.

💬 Posted 2 inline comments

File: ultralytics/utils/torch_utils.py [hunk 1]
```diff
@@ -106,10 +106,14 @@ def autocast(enabled: bool, device: str = "cuda"):
 
     Notes:
         - For PyTorch versions 1.13 and newer, it uses `torch.amp.autocast`.
-        - For older versions, it uses `torch.cuda.amp.autocast`.
+        - For older versions, it uses the backend-specific AMP context.
     """
     if TORCH_1_13:
         return torch.amp.autocast(device, enabled=enabled)
+    elif device == "npu":
```
Comment (glenn-jocher) [inline ultralytics/utils/torch_utils.py]: Fixed in 0d67aebea: the NPU-specific torch_npu autocast path now takes precedence on every PyTorch version, and the regression test explicitly exercises the TORCH_1_13+ branch.

File: ultralytics/engine/validator.py [hunk 2]
```diff
@@ -179,10 +179,13 @@ def __call__(self, trainer=None, model=None):
                 self.args.data = convert_ndjson_to_yolo_if_needed(self.args.data)
             model = AutoBackend(
                 model=model or self.args.model,
-                # DDP ranks reuse the device assigned in trainer._setup_ddp() via torch.cuda.set_device()
+                # DDP ranks reuse the device assigned in trainer._setup_ddp()
                 device=select_device(self.args.device)
                 if RANK == -1
-                else torch.device("cuda", torch.cuda.current_device()),
+                else torch.device(
+                    "npu" if self.args.device.startswith("npu") else "cuda",
```
Comment (glenn-jocher) [inline ultralytics/engine/validator.py]: Fixed in 0d67aebea: Profile now synchronizes torch.npu for validation timing, BaseModel layer profiling passes the model device to time_sync(), and focused NPU synchronization coverage was added.

**Review by UltralyticsAssistant — DISMISSED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

No genuine issues found. The changes consistently extend the existing device, DDP, AMP, memory, profiling, AutoBatch, EMA, and batch-transfer owners for Ascend NPUs while preserving CUDA paths. The added tests cover the key parsing, rank binding, HCCL selection, autocast, synchronization, and clipping behavior.

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

The NPU device selection, HCCL DDP setup, AMP routing, validation device reuse, and profiling integration are consistently wired through existing owners. One AutoBatch accounting issue will make the newly supported NPU AutoBatch unnecessarily conservative.

💬 Posted 1 inline comment

File: ultralytics/utils/autobatch.py [hunk 3]
```diff
@@ -70,22 +70,25 @@ def autobatch(
...
+    accelerator = torch.npu if device.type == "npu" else torch.cuda
     gb = 1 << 30  # bytes to GiB (1024 ** 3)
-    d = f"CUDA:{device.index}"  # 'CUDA:0'
-    properties = torch.cuda.get_device_properties(device)  # device properties
+    d = f"{device.type.upper()}:{device.index}"
+    properties = accelerator.get_device_properties(device)  # device properties
     t = properties.total_memory / gb  # GiB total
-    r = torch.cuda.memory_reserved(device) / gb  # GiB reserved
-    a = torch.cuda.memory_allocated(device) / gb  # GiB allocated
+    r = accelerator.memory_reserved(device) / gb  # GiB reserved
```
Comment (glenn-jocher) [inline ultralytics/utils/autobatch.py:88]: Fixed at the shared AutoBatch owner in 0b668a8fc: free memory is now total minus reserved, and predicted utilization adds reserved exactly once. Allocated memory remains separately logged but is no longer double-counted for CUDA or NPU.

File: ultralytics/models/yolo/detect/train.py [hunk 4]
```diff
@@ -114,7 +114,7 @@ def preprocess_batch(self, batch: dict) -> dict:
         """
         for k, v in batch.items():
             if isinstance(v, torch.Tensor):
-                batch[k] = v.to(self.device, non_blocking=self.device.type == "cuda")
+                batch[k] = v.to(self.device, non_blocking=self.device.type in {"cuda", "npu"})
```
Comment (UltralyticsAssistant) [inline ultralytics/models/yolo/detect/train.py]: 💡 **MEDIUM**: NPU batches are still built without pinned host memory: `build_dataloader()` gates `pin_memory` (and its worker-count division) exclusively on `torch.cuda.device_count()`, which is zero on an NPU-only host. Thus these new `non_blocking=True` NPU transfers do not provide the intended asynchronous input pipeline, and multi-NPU runs can create too many workers. Make the dataloader's accelerator detection account for `torch.npu` as well.

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

The device parsing, rank-local NPU binding, HCCL backend selection, AMP, memory, EMA, profiling, and validation changes are integrated into the existing owners cleanly. One input-pipeline gap prevents the newly enabled NPU non-blocking transfers from being effective.

💬 Posted 1 inline comment

File: ultralytics/models/yolo/detect/train.py (hunk 4 above)
Comment (glenn-jocher) [reply ultralytics/models/yolo/detect/train.py]: Fixed in 9b2efe429 at the dataloader owner: build_dataloader now selects the available NPU backend before device_count(), enabling pinned memory and per-device worker division on NPU-only hosts. Added focused hardware-free coverage, and consolidated the same backend-selection pattern across the touched trainer, validator, memory, and profiling owners.

**Comment by UltralyticsAssistant** [comment]
🎉 Merged—thank you, @glenn-jocher, for this substantial expansion of Ultralytics hardware support!

> “Great things are done by a series of small things brought together.” — Vincent van Gogh

This PR brings together ROCm, Ascend NPU, Intel XPU, and CUDA-aware improvements across training, inference, distributed execution, AMP, profiling, and memory management. By making Ultralytics more device-agnostic, it opens YOLO workflows to a broader range of hardware while preserving existing CUDA support. Your work meaningfully advances accessible, reliable AI acceleration for the community—thank you!
